src/stream/futures.rs

```diff
@@ -0,0 +1,67 @@
+use future::{Future, IntoFuture};
+use stream::Stream;
+use poll::Poll;
+use Async;
+
+use std::prelude::v1::*;
+
+/// An adaptor for a stream of futures to execute the futures concurrently, if
+/// possible, delivering results as they become available.
+///
+/// This adaptor will return their results in the order that they complete.
+/// This is created by the `futures` method.
+///
+#[must_use = "streams do nothing unless polled"]
+pub struct Futures<F>
+    where F: Future
+{
+    futures: Vec<F>,
+}
+
+/// Converts a `Vec` of `IntoFuture` elements into a `Stream` over the underlying
+/// futures. The created stream will yield into values produced by those futures.
+/// Similar to `buffer_unordered` the order of the produced items will not be equal
+/// with the order of elements in the input vector.
+///
+pub fn futures<I, It, F>(futures: I) -> Futures<F::Future>
+    where I: IntoIterator<Item = F, IntoIter = It>,
+          It: Iterator<Item = F> + ExactSizeIterator,
+          F: IntoFuture
+{
+    Futures {
+        futures: futures.into_iter().map(IntoFuture::into_future).collect(),
+    }
+}
+
+
+impl<F> Stream for Futures<F>
+    where F: Future
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.futures.is_empty() {
+            return Ok(Async::Ready(None));
+        }
+        let mut ready = None;
+        for (id, f) in self.futures.iter_mut().enumerate() {
+            match f.poll() {
+                Ok(Async::NotReady) => {}
+                Ok(Async::Ready(val)) => {
+                    ready = Some((id, val));
+                    break;
+                }
+                Err(e) => {
+                    return Err(e);
```

Comment:
> What is the problem with a stream returning the same error when polled again?

Polling an unfused future after it has completed is an error, from `Future::poll`:

> Once a future has completed (returned Ready or Err from poll), then any future calls to poll may panic, block forever, or otherwise cause wrong behavior. The Future trait itself provides no guarantees about the behavior of poll after a future has completed.

> I think if an error occurred the whole stream should treated as failed.

I disagree, its useful to let the application decide this instead of the implementation.